src/common/range.rs

```diff
@@ -141,278 +146,454 @@ impl ByteRangeSpec {
-            Range::Bytes(ref ranges) => {
-                try!(write!(f, "bytes="));
+    /// Add a range to the header. Fails if the end bound of the range
+    /// is before the start bound.
+    pub fn range(mut self, bounds: impl RangeBounds<u64>) -> Result<Self, InvalidRange> {
+        match reduce_bounds(bounds) {
+            (first, Some(last)) => {
+                if first > last {
+                    return Err(err());
+                }
+                self.inner.push(ByteRangeSpec::FromTo(first, last));
+            }
+            (first, None) => {
+                self.inner.push(ByteRangeSpec::AllFrom(first));
+            }
+        };
+        Ok(self)
+    }
+
+    /// Add a suffix to the header. Fails if the suffix length is zero.
+    pub fn suffix(mut self, len: u64) -> Result<Self, InvalidRange> {
+        let len = match NonZeroU64::new(len) {
+            Some(len) => len,
+            None => return Err(err()),
+        };
+
+        self.inner.push(ByteRangeSpec::Last(len));
+        Ok(self)
+    }
+
+    /// Finish up the header. Fails if no ranges or suffixes were added.
+    pub fn finish(self) -> Result<Range, InvalidRange> {
+        if self.inner.is_empty() {
+            return Err(err());
+        }
+        Ok(Range::Bytes(self.inner))
+    }
+}
 
+impl std::fmt::Display for Range {
+    fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
+        match self {
+            Range::Bytes(ranges) => {
+                write!(f, "bytes=")?;
                 for (i, range) in ranges.iter().enumerate() {
                     if i != 0 {
-                        try!(f.write_str(","));
+                        f.write_str(",")?;
                     }
-                    try!(Display::fmt(range, f));
+                    std::fmt::Display::fmt(range, f)?;
                 }
                 Ok(())
-            },
-            Range::Unregistered(ref unit, ref range_str) => {
-                write!(f, "{}={}", unit, range_str)
-            },
+            }
+            Range::Unregistered { unit, set } => write!(f, "{}={}", unit, set),
         }
     }
 }
 
 impl FromStr for Range {
-    type Err = ::Error;
+    type Err = InvalidRange;
 
-    fn from_str(s: &str) -> ::Result<Range> {
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
         let mut iter = s.splitn(2, '=');
 
         match (iter.next(), iter.next()) {
-            (Some("bytes"), Some(ranges)) => {
-                let ranges = from_comma_delimited(ranges);
-                if ranges.is_empty() {
-                    return Err(::Error::Header);
+            (Some("bytes"), Some(set)) => {
+                let mut invalid = false;
+                let specs: Vec<ByteRangeSpec> = set
+                    .split(',')
+                    .filter_map(|spec: &str| -> Option<ByteRangeSpec> {
+                        let spec = spec.parse();
+                        if spec.is_err() {
+                            invalid = true;
```

Comment:
My main problem here is that we don't have any way to communicate which bit we stumbled over back to a dev that might be debugging this, without them actually connecting a debugger and stepping down this far. If we were using `log`, we'd be able to put out debug logs here, which I'd be happier with. Would a `log` feature that adds a dependency to `log` be acceptable here, so that we can log this?

Reply:
Logs are nice, but not super discoverable or obvious (even if documented) compared to actual errors. Really the best thing would be to have a useful error type that the caller can inspect.